extension/src/state_aggregate.rs

```diff
@@ -262,7 +352,89 @@ impl toolkit_experimental::state_agg {
                     state_end,
                 });
             }
-            StateAgg::new(states, durations, first, last)
+            StateAgg::new(states, durations, first, last, None)
+        })
+    }
+}
+
+#[aggregate]
+impl toolkit_experimental::timeline_agg {
+    type State = StateAggTransState;
+
+    const PARALLEL_SAFE: bool = true;
+
+    fn transition(
+        state: Option<State>,
+        #[sql_type("timestamptz")] ts: TimestampTz,
+        #[sql_type("text")] value: Option<String>,
+    ) -> Option<State> {
+        state_agg::transition(state, ts, value)
+    }
+
+    fn combine(a: Option<&State>, b: Option<&State>) -> Option<State> {
+        state_agg::combine(a, b)
+    }
+
+    fn serialize(state: &mut State) -> bytea {
+        state_agg::serialize(state)
+    }
+
+    fn deserialize(bytes: bytea) -> State {
+        state_agg::deserialize(bytes)
+    }
+
+    fn finally(state: Option<&mut State>) -> Option<TimelineAgg<'static>> {
+        state.map(|s| {
+            let mut states = String::new();
+            let mut durations: Vec<DurationInState> = vec![];
+            let (map, first, last) = s.make_duration_map_and_bounds();
+            for (state, duration) in map {
+                let state_beg = states.len() as u32;
+                let state_end = state_beg + state.len() as u32;
+                states.push_str(&state);
+                durations.push(DurationInState {
+                    duration,
+                    state_beg,
+                    state_end,
+                });
+            }
+
+            let mut merged_durations: Vec<TimeInState> = Vec::new();
+            let mut last_record_state = None;
+            for record in s.records.drain(..) {
+                let state_beg = states
```

Comment:
In #640 I reworked the way states are stored. After that PR it will store "on" and "powering_on" with a state string of just `powering_on` where "powering_on" is stored as `(0, 11)` and "on" is stored as `(9, 11)`. The code can already handle overlapping state names.